File: frontend/lib/api/digitalTwin.ts
import { api } from './client';

export interface DigitalTwinOverview {
  status: string;
  tenant_id: string;
  twin_models_count: number;
  composite_health_score: number;
  latest_snapshot_code: string;
  latest_snapshot_timestamp: string;
  state_hash: string;
  commercial_state: Record<string, any>;
  financial_state: Record<string, any>;
  delivery_state: Record<string, any>;
  ai_state: Record<string, any>;
  reliability_state: Record<string, any>;
  risk_state: Record<string, any>;
}

export interface StateSnapshot {
  id?: string;
  snapshot_code: string;
  tenant_id: string;
  title: string;
  snapshot_timestamp: string;
  commercial_state: Record<string, any>;
  delivery_state: Record<string, any>;
  operational_state: Record<string, any>;
  financial_state: Record<string, any>;
  ai_state: Record<string, any>;
  reliability_state: Record<string, any>;
  risk_state: Record<string, any>;
  composite_health_score: number;
  state_hash: string;
  is_valid: boolean;
}

export interface Scenario {
  id?: string;
  scenario_code: string;
  tenant_id: string;
  twin_model_id: string;
  name: string;
  description?: string;
  scenario_type: string;
  time_horizon: string;
  simulation_method: string;
  parameter_overrides: Record<string, number>;
  assumptions: Array<{
    name: string;
    parameter_code: string;
    baseline_value: number;
    assumed_value: number;
    delta_percentage: number;
  }>;
  constraints: Array<Record<string, any>>;
  status: string;
  created_by: string;
  version: number;
}

export interface SimulationResult {
  simulation_code: string;
  scenario_id: string;
  method: string;
  iterations: number;
  runtime_seconds: number;
  metrics_summary: {
    horizon_months?: number;
    cumulative_revenue_usd?: number;
    cumulative_gross_profit_usd?: number;
    cumulative_net_profit_usd?: number;
    cumulative_ai_cost_usd?: number;
    average_gross_margin_percentage?: number;
    ending_active_clients?: number;
    final_capacity_utilization_percentage?: number;
  };
  uncertainty_distribution: Record<string, {
    expected: number;
    p10: number;
    p25: number;
    p50: number;
    p75: number;
    p90: number;
    min: number;
    max: number;
  }>;
  constraint_violations: string[];
  is_sandboxed: boolean;
  assumptions_applied: Array<{ code: string; name: string; delta: number }>;
}

export interface ScenarioComparison {
  baseline: {
    scenario_name: string;
    scenario_code: string;
    cumulative_revenue_usd: number;
    cumulative_net_profit_usd: number;
    capacity_utilization_percentage: number;
    simulation_result: SimulationResult;
  };
  alternatives: Array<{
    scenario_name: string;
    scenario_code: string;
    method: string;
    cumulative_revenue_usd: number;
    revenue_delta_usd: number;
    revenue_delta_percentage: number;
    cumulative_net_profit_usd: number;
    net_profit_delta_usd: number;
    capacity_utilization_percentage: number;
    constraint_violations: string[];
    simulation_result: SimulationResult;
  }>;
}

export interface SensitivityRanking {
  parameter_code: string;
  parameter_name: string;
  target_metric: string;
  sensitivity_score: number;
  impact_level: string;
  low_impact_value: number;
  high_impact_value: number;
}

export interface CounterfactualResult {
  counterfactual_code: string;
  historical_event_id: string;
  hypothetical_condition: string;
  historical_actual: Record<string, any>;
  simulated_alternative: Record<string, any>;
  divergence_summary: string;
  limitations: string;
}

export interface DecisionOption {
  id?: string;
  option_code: string;
  title: string;
  scenario_id: string;
  expected_benefit_usd: number;
  expected_cost_usd: number;
  risk_score: number;
  confidence_score: number;
  tradeoff_summary: string;
}

export interface DecisionRecord {
  id?: string;
  decision_code: string;
  question: string;
  selected_option_id?: string;
  rationale: string;
  decision_owner: string;
  approved_at: string;
  scenario_version: number;
  status: string;
}

export interface OutcomeRecord {
  id?: string;
  decision_id?: string;
  scenario_id?: string;
  observed_period: string;
  predicted_metrics: Record<string, any>;
  actual_metrics: Record<string, any>;
  variance_percentage: number;
  model_error: number;
}

export interface CalibrationReport {
  calibrations_count: number;
  parameter_updates: Array<{
    parameter_code: string;
    recommended_action: string;
    reason: string;
  }>;
  average_prediction_error: number;
  summary: string;
}

export interface CopilotResponse {
  query: string;
  intent: string;
  scenario_generated?: Record<string, any>;
  simulation_result?: SimulationResult;
  explanation: string;
  confidence: number;
  governance_notice: string;
}

export const digitalTwinApi = {
  getOverview: (tenantId: string = 'default_tenant') =>
    api.get<DigitalTwinOverview>('/api/v1/digital-twin/overview', { params: { tenant_id: tenantId } }),

  getState: (tenantId: string = 'default_tenant') =>
    api.get<StateSnapshot>('/api/v1/digital-twin/state', { params: { tenant_id: tenantId } }),

  createSnapshot: (title: string, tenantId: string = 'default_tenant') =>
    api.post<StateSnapshot>('/api/v1/digital-twin/state/snapshot', { title }, { params: { tenant_id: tenantId } }),

  getParameters: () =>
    api.get<{ status: string; parameters: Record<string, any> }>('/api/v1/digital-twin/parameters'),

  createScenario: (payload: {
    name: string;
    scenario_type?: string;
    time_horizon?: string;
    simulation_method?: string;
    parameter_overrides?: Record<string, number>;
  }, tenantId: string = 'default_tenant') =>
    api.post<Scenario>('/api/v1/digital-twin/scenarios', payload, { params: { tenant_id: tenantId } }),

  runSimulation: (payload: {
    scenario_name?: string;
    parameter_overrides?: Record<string, number>;
    simulation_method?: string;
    time_horizon?: string;
    iterations?: number;
    random_seed?: number;
  }, tenantId: string = 'default_tenant') =>
    api.post<SimulationResult>('/api/v1/digital-twin/simulations/run', payload, { params: { tenant_id: tenantId } }),

  compareScenarios: (payload: {
    baseline_scenario_name?: string;
    baseline_overrides?: Record<string, number>;
    scenarios: Array<{ name: string; parameter_overrides: Record<string, number> }>;
    iterations?: number;
  }, tenantId: string = 'default_tenant') =>
    api.post<ScenarioComparison>('/api/v1/digital-twin/compare', payload, { params: { tenant_id: tenantId } }),

  getSensitivity: (payload: {
    target_metric?: string;
    parameter_overrides?: Record<string, number>;
    sweep_percentage?: number;
  }, tenantId: string = 'default_tenant') =>
    api.post<SensitivityRanking[]>('/api/v1/digital-twin/sensitivity', payload, { params: { tenant_id: tenantId } }),

  runCounterfactual: (payload: {
    historical_event_id: string;
    hypothetical_condition: string;
    historical_actual_metrics: Record<string, any>;
    hypothetical_parameter_overrides: Record<string, number>;
  }) =>
    api.post<CounterfactualResult>('/api/v1/digital-twin/counterfactuals', payload),

  recordDecision: (payload: {
    question: string;
    rationale: string;
    decision_owner: string;
    selected_option_id?: string;
  }) =>
    api.post<DecisionRecord>('/api/v1/digital-twin/decisions', payload),

  recordOutcome: (payload: {
    observed_period: string;
    predicted_metrics: Record<string, any>;
    actual_metrics: Record<string, any>;
    decision_id?: string;
    scenario_id?: string;
  }) =>
    api.post<OutcomeRecord>('/api/v1/digital-twin/outcomes', payload),

  runCalibration: (outcomes: Array<{
    observed_period: string;
    predicted_metrics: Record<string, any>;
    actual_metrics: Record<string, any>;
  }>) =>
    api.post<CalibrationReport>('/api/v1/digital-twin/calibration', outcomes),

  queryCopilot: (query: string, tenantId: string = 'default_tenant') =>
    api.post<CopilotResponse>('/api/v1/digital-twin/copilot', { query, tenant_id: tenantId }),
};

File: backend/app/schemas/digital_twin.py
"""
Pydantic Schemas for Phase 50:
Unified Digital Twin, Business Simulation, Scenario Intelligence & Strategic What-If Engine.
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class StateSnapshotCreateRequest(BaseModel):
    title: str = "Enterprise Point-in-Time Snapshot"
    source_snapshot_id: Optional[str] = None


class StateSnapshotResponse(BaseModel):
    id: Optional[str] = None
    snapshot_code: str
    tenant_id: str
    title: str
    snapshot_timestamp: datetime
    commercial_state: Dict[str, Any] = Field(default_factory=dict)
    delivery_state: Dict[str, Any] = Field(default_factory=dict)
    operational_state: Dict[str, Any] = Field(default_factory=dict)
    financial_state: Dict[str, Any] = Field(default_factory=dict)
    ai_state: Dict[str, Any] = Field(default_factory=dict)
    reliability_state: Dict[str, Any] = Field(default_factory=dict)
    risk_state: Dict[str, Any] = Field(default_factory=dict)
    composite_health_score: float
    state_hash: str
    is_valid: bool = True


class ScenarioCreateRequest(BaseModel):
    name: str
    twin_model_id: str = "primary_twin_model"
    scenario_type: str = "GROWTH"
    time_horizon: str = "12_MONTHS"
    simulation_method: str = "MONTE_CARLO"
    parameter_overrides: Dict[str, float] = Field(default_factory=dict)
    custom_assumptions: List[Dict[str, Any]] = Field(default_factory=list)
    constraints: List[Dict[str, Any]] = Field(default_factory=list)


class ScenarioResponse(BaseModel):
    id: Optional[str] = None
    scenario_code: str
    tenant_id: str
    twin_model_id: str
    name: str
    description: Optional[str] = None
    scenario_type: str
    time_horizon: str
    simulation_method: str
    parameter_overrides: Dict[str, float] = Field(default_factory=dict)
    assumptions: List[Dict[str, Any]] = Field(default_factory=list)
    constraints: List[Dict[str, Any]] = Field(default_factory=list)
    status: str
    created_by: str
    version: int


class SimulationRunRequest(BaseModel):
    scenario_name: Optional[str] = None
    parameter_overrides: Dict[str, float] = Field(default_factory=dict)
    simulation_method: str = "MONTE_CARLO"
    time_horizon: str = "12_MONTHS"
    iterations: int = 1000
    random_seed: Optional[int] = 42


class SimulationResponse(BaseModel):
    simulation_code: str
    scenario_id: str
    method: str
    iterations: int
    runtime_seconds: float
    metrics_summary: Dict[str, Any] = Field(default_factory=dict)
    uncertainty_distribution: Dict[str, Any] = Field(default_factory=dict)
    constraint_violations: List[str] = Field(default_factory=list)
    is_sandboxed: bool = True
    assumptions_applied: List[Dict[str, Any]] = Field(default_factory=list)


class ScenarioCompareRequest(BaseModel):
    baseline_scenario_name: str = "Current Baseline"
    baseline_overrides: Dict[str, float] = Field(default_factory=dict)
    scenarios: List[Dict[str, Any]] = Field(default_factory=list)
    iterations: int = 1000


class ScenarioCompareResponse(BaseModel):
    baseline: Dict[str, Any]
    alternatives: List[Dict[str, Any]]


class SensitivityRequest(BaseModel):
    target_metric: str = "cumulative_revenue_usd"
    parameter_overrides: Dict[str, float] = Field(default_factory=dict)
    sweep_percentage: float = 0.25


class SensitivityRankingResponse(BaseModel):
    parameter_code: str
    parameter_name: str
    target_metric: str
    sensitivity_score: float
    impact_level: str
    low_impact_value: float
    high_impact_value: float


class CounterfactualRequest(BaseModel):
    historical_event_id: str
    hypothetical_condition: str
    historical_actual_metrics: Dict[str, Any]
    hypothetical_parameter_overrides: Dict[str, float]


class CounterfactualResponse(BaseModel):
    counterfactual_code: str
    historical_event_id: str
    hypothetical_condition: str
    historical_actual: Dict[str, Any]
    simulated_alternative: Dict[str, Any]
    divergence_summary: str
    limitations: str


class DecisionOptionResponse(BaseModel):
    id: Optional[str] = None
    option_code: str
    title: str
    scenario_id: str
    expected_benefit_usd: float
    expected_cost_usd: float
    risk_score: float
    confidence_score: float
    tradeoff_summary: str


class DecisionRecordCreateRequest(BaseModel):
    question: str
    rationale: str
    decision_owner: str
    selected_option_id: Optional[str] = None


class DecisionRecordResponse(BaseModel):
    id: Optional[str] = None
    decision_code: str
    question: str
    selected_option_id: Optional[str] = None
    rationale: str
    decision_owner: str
    approved_at: datetime
    scenario_version: int
    status: str


class OutcomeRecordCreateRequest(BaseModel):
    observed_period: str
    predicted_metrics: Dict[str, Any]
    actual_metrics: Dict[str, Any]
    decision_id: Optional[str] = None
    scenario_id: Optional[str] = None


class OutcomeRecordResponse(BaseModel):
    id: Optional[str] = None
    decision_id: Optional[str] = None
    scenario_id: Optional[str] = None
    observed_period: str
    predicted_metrics: Dict[str, Any]
    actual_metrics: Dict[str, Any]
    variance_percentage: float
    model_error: float


class CalibrationReportResponse(BaseModel):
    calibrations_count: int
    parameter_updates: List[Dict[str, Any]] = Field(default_factory=list)
    average_prediction_error: float
    summary: str


class CopilotQueryRequest(BaseModel):
    query: str
    tenant_id: str = "default_tenant"


class CopilotQueryResponse(BaseModel):
    query: str
    intent: str
    scenario_generated: Optional[Dict[str, Any]] = None
    simulation_result: Optional[Dict[str, Any]] = None
    explanation: str
    confidence: float
    governance_notice: str = "Simulations are estimates. Strategic decisions require human executive signoff."
